### ISSUE TITLE
After tokens refresh first time, every subsequent protected resource request results in token refreshes

### ISSUE THREAD

**Issue by mikebper** [AzureAD/microsoft-authentication-library-for-js#5900]
### Core Library

MSAL.js v2 (@azure/msal-browser)

### Core Library Version

2.31.0

### Wrapper Library

MSAL Angular (@azure/msal-angular)

### Wrapper Library Version

2.4.6

### Public or Confidential Client?

Public

### Description

I am seeing an issue where the following occurs:

1. Tokens are refreshed by aquireTokenSilent using the refresh token, which happens at the first request for a protected resource that is less than 5 minutes from the id/access token expirations.
2. New ID, access, and refresh tokens are issued, which all look good. The ID and access tokens have full lifetimes, and the refresh token lifetime is the remainder of the day (1440 minutes) from login.
3. After the first time the tokens are refreshed, EVERY subsequent request for a protected resource results in a call to the token endpoint to refresh the tokens again, even though it is not less than 5 minutes from the previously acquired tokens' expirations.

I tried using both BrowserCacheLocation.SessionStorage and BrowserCacheLocation.LocalStorage and the same thing happens with both. I've been testing with my token lifetimes set to 5 minutes and 20 seconds, so that they refresh every 20 seconds. I changed that to 10 minutes so that the first refresh doesn't occur for 5 minutes, and the same problem occurs. I've done this in multiple browsers with cleared cache.

I think the issue may be related to the fact that I also do not see the new tokens in storage when I check in the browsers' dev tools, despite the updated ID token reflecting in the active account's idTokenClaims, and the updated access token being the one added by MSAL to protected resource requests.

### Error Message

n/a

### Msal Logs

n/a

### MSAL Configuration

```javascript
app.module.ts:
MsalModule.forRoot(new PublicClientApplication(msalConfig),
    {
      interactionType: InteractionType.Redirect,
      authRequest: {
        scopes: protectedResource.authRequest.scopes,
        redirectUri: '/'
      }
    },
    {
      interactionType: InteractionType.Redirect,
      protectedResourceMap: new Map([
        [protectedResource.myApiEndpoint.endpoint, protectedResource.myApiEndpoint.scopes],
      ])
    })
],
providers: [    
  {
    provide: HTTP_INTERCEPTORS,
    useClass: MsalInterceptor,
    multi: true
  },
  MsalGuard
],
bootstrap: [
  AppComponent,
  MsalRedirectComponent
]

auth-config.ts:
import { LogLevel, Configuration, BrowserCacheLocation } from '@azure/msal-browser';
import { environment } from 'src/environments/environment';

export const b2cPolicies = {
    authorities: {
        signIn: {
            authority: 'https://myapib2cdev.b2clogin.com/myapib2cdev.onmicrosoft.com/B2C_1A_SIGNIN',
        },
        resetPassword: {
            authority: 'https://myapib2cdev.b2clogin.com/myapib2cdev.onmicrosoft.com/B2C_1A_PASSWORDRESET',
        },
        editProfile: {
            authority: 'https://myapib2cdev.b2clogin.com/myapib2cdev.onmicrosoft.com/B2C_1A_PROFILEEDIT'
        }
    },
    authorityDomain: 'myapib2cdev.b2clogin.com'
};

export const msalConfig: Configuration = {
    auth: {
        clientId: 'my-client-ID',
        authority: b2cPolicies.authorities.signIn.authority,
        knownAuthorities: [b2cPolicies.authorityDomain],
        redirectUri: '/',
        postLogoutRedirectUri: '/'
    },
    cache: {
        cacheLocation: BrowserCacheLocation.SessionStorage
    },
    system: {
        loggerOptions: {
            logLevel: LogLevel.Error,
            loggerCallback: (level, message, containsPii) => {
                if (containsPii) {
                    return;
                }
                switch (level) {
                    case LogLevel.Error:
                        console.error(message);
                        return;
                    case LogLevel.Info:
                        console.info(message);
                        return;
                    case LogLevel.Verbose:
                        console.debug(message);
                        return;
                    case LogLevel.Warning:
                        console.warn(message);
                        return;
                }
            },
            piiLoggingEnabled: false
        }
    }
}

export const protectedResource = {
    authRequest: {
        scopes: ['https://myapib2cdev.onmicrosoft.com/myapi/Read']
    },
    myApiEndpoint: {
        endpoint: `${environment.myApiBaseUrl}/myApiEndpoint`,
        scopes: ['https://myapib2cdev.onmicrosoft.com/myapi/Read']
    },
}
```

### Relevant Code Snippets

```javascript
n/a
```

### Reproduction Steps

1. Log in
2. Check local storage or session storage in the browser dev tools and note the ID, access, and refresh token values.
3. Wait until it's less than 5 minutes from the ID/Access token expirations and access a protected resource.
4. Observe in dev tools Network section that the token endpoint is called first and new tokens are provided in the response.
5. Check local storage or session storage in the browser dev tools and observe that the old tokens from step 2 are still there.
6. Debug to check that the active accout (getActiveAcount()) AccountInfo idTokenClaims do reflect the new token despite it not showing in the browser's storage in dev tools.
7. Confirm at the protected resource that the new access token is passed despite it not showing in the brower's storage in dev tools.
8. Access a protected resource prior to the newly acquired tokens (from step 4) expiring, and observe in dev tools Network section that the token endpoint is called again when it should not be.
9. Make additional requests and note that for every one the token endpoint is called and new tokens are provided in response.

### Expected Behavior

New tokens are only acquired less than 5 minutes from token expiration. When new tokens are acquired, local storage or session storage are updated accordingly.

### Identity Provider

Azure B2C Custom Policy

### Browsers Affected (Select all that apply)

Chrome, Firefox, Edge

### Source

External (Customer)

**Comment by mikebper** [comment]
I did some more testing. The app in question is using AAD B2C as the IDP. When logging in with a local AAD B2C account the problem does NOT happen. A regular AAD (not B2C) directory is set up as a social login option, and only when logging in like that does the problem occur. I'm not sure if it's something wrong with the app or the B2C setup, or MSAL.

**Comment by mikebper** [comment]
The problem does appear to be MSAL to me. When logging in as the local B2C account, the tokens are updated in session storage (visible in dev tools), and subsequent requests don't always get new tokens--only when <5 minutes from expiration.

When logging in as the AAD account as a "social login", the tokens returned from the refresh request are correct, but are not updated in session storage, and every subsequent request refreshes them again, even though it's not <5 minutes from their expiration.

**Comment by mikebper** [comment]
I compared the responses from the refresh token requests for the local B2C user where it works, and the AAD social login user where it doesn't, and they look comparable.

**Comment by mikebper** [comment]
This is the relevant code from my app where MSAL is being used:

```typescript
  ngOnInit(): void {
    this.msalBroadcastService.msalSubject$
      .pipe(
        filter((msg: EventMessage) => msg.eventType === EventType.ACQUIRE_TOKEN_SUCCESS),
        takeUntil(this.destroying)
      )
      .subscribe((result: EventMessage) => {
        const payload = result.payload as { account: AccountInfo };
        this.updateAccount(payload.account)
      });

    this.msalBroadcastService.inProgress$
      .pipe(
        filter((status: InteractionStatus) => status === InteractionStatus.None),
        takeUntil(this.destroying)
      )
      .subscribe(() => {
        let activeAccount = this.msalService.instance.getActiveAccount();
        if (!activeAccount) {
          let accounts = this.msalService.instance.getAllAccounts();
          if (accounts.length > 0) {
            activeAccount = accounts[0];
            this.msalService.instance.setActiveAccount(activeAccount);
            this.setLoginDisplay(activeAccount);
          }
        } else {
          this.setLoginDisplay(activeAccount);
        }
      });
  }

  setLoginDisplay(activeAccount: AccountInfo): void {
    this.updateAccount(activeAccount);
    this.isAuthenticated = true;
    ...update the UI
  }

  updateAccount(activeAccount: AccountInfo) {
    this.authService.account = activeAccount;
    this.displayName = `${activeAccount.idTokenClaims!['given_name']} ${activeAccount.idTokenClaims!['family_name']}`;
  }
```

**Comment by SamP98** [comment]
We are experiencing the same issue on AAD B2C. Comparing our setup to yours, its the exact same. We have it set up following this documentation https://learn.microsoft.com/en-us/azure/active-directory-b2c/enable-authentication-angular-spa-app

@angular/core: 15.2.0
@azure/msal-angular: 2.5.4
@azure/msal-browser: 2.34.0

For us, it doesn't matter how old or new the token is, it could be a brand new token but the MSAL interceptor is triggering aquireTokenSilent whenever we try to make an http request to our api.

Here are our logs. MSAL recognizes that an active account is logged in and hits the cache, and yet refreshes anyways despite the token being good for another hour.

`[Wed, 19 Apr 2023 19:48:43 GMT] : @azure/msal-angular@2.5.4 : Verbose - MSAL Interceptor activated
auth-config.ts:10 [Wed, 19 Apr 2023 19:48:43 GMT] : @azure/msal-angular@2.5.4 : Verbose - Interceptor - getting scopes for endpoint
auth-config.ts:10 [Wed, 19 Apr 2023 19:48:43 GMT] : @azure/msal-browser@2.34.0 : Trace - BrowserCacheManager.getActiveAccount: Active account filters schema found
auth-config.ts:10 [Wed, 19 Apr 2023 19:48:43 GMT] : @azure/msal-browser@2.34.0 : Trace - BrowserCacheManager.getAccessTokenCredential: cache hit
auth-config.ts:10 [Wed, 19 Apr 2023 19:48:43 GMT] : @azure/msal-browser@2.34.0 : Trace - BrowserCacheManager.getIdTokenCredential: cache hit
auth-config.ts:10 [Wed, 19 Apr 2023 19:48:43 GMT] : @azure/msal-browser@2.34.0 : Trace - BrowserCacheManager.getRefreshTokenCredential: cache hit
auth-config.ts:10 [Wed, 19 Apr 2023 19:48:43 GMT] : @azure/msal-browser@2.34.0 : Trace - BrowserCacheManager.getAccountInfoByFilter: total 1 accounts found
auth-config.ts:10 [Wed, 19 Apr 2023 19:48:43 GMT] : @azure/msal-angular@2.5.4 : Verbose - Interceptor - active account selected
auth-config.ts:10 [Wed, 19 Apr 2023 19:48:43 GMT] : @azure/msal-browser@2.34.0 : Trace - BrowserCacheManager.getActiveAccount: Active account filters schema found
auth-config.ts:10 [Wed, 19 Apr 2023 19:48:43 GMT] : @azure/msal-browser@2.34.0 : Trace - BrowserCacheManager.getAccessTokenCredential: cache hit
auth-config.ts:10 [Wed, 19 Apr 2023 19:48:43 GMT] : @azure/msal-browser@2.34.0 : Trace - BrowserCacheManager.getIdTokenCredential: cache hit
auth-config.ts:10 [Wed, 19 Apr 2023 19:48:43 GMT] : @azure/msal-browser@2.34.0 : Trace - BrowserCacheManager.getRefreshTokenCredential: cache hit
auth-config.ts:10 [Wed, 19 Apr 2023 19:48:43 GMT] : @azure/msal-browser@2.34.0 : Trace - BrowserCacheManager.getAccountInfoByFilter: total 1 accounts found
auth-config.ts:10 [Wed, 19 Apr 2023 19:48:43 GMT] : @azure/msal-angular@2.5.4 : Info - Interceptor - 1 scopes found for endpoint
auth-config.ts:10 [Wed, 19 Apr 2023 19:48:43 GMT] : @azure/msal-angular@2.5.4 : Info - Interceptor - [https://MYB2C.onmicrosoft.com/api/tasks.read] scopes found for https://OURDEVAPI.azurewebsites.net/api/APICALLHERE
auth-config.ts:10 [Wed, 19 Apr 2023 19:48:43 GMT] : [d665fe96-d678-42e8-94a0-5d84bf35dadf] : @azure/msal-browser@2.34.0 : Trace - PerformanceClient: Performance measurement started for acquireTokenSilent
auth-config.ts:10 [Wed, 19 Apr 2023 19:48:43 GMT] : [d665fe96-d678-42e8-94a0-5d84bf35dadf] : @azure/msal-browser@2.34.0 : Trace - PerformanceClient: Performance measurement for acquireTokenSilent started
auth-config.ts:10 [Wed, 19 Apr 2023 19:48:43 GMT] : @azure/msal-browser@2.34.0 : Trace - PerformanceClient: Updating static fields
auth-config.ts:10 [Wed, 19 Apr 2023 19:48:43 GMT] : @azure/msal-browser@2.34.0 : Verbose - preflightBrowserEnvironmentCheck started
auth-config.ts:10 [Wed, 19 Apr 2023 19:48:43 GMT] : [d665fe96-d678-42e8-94a0-5d84bf35dadf] : @azure/msal-browser@2.34.0 : Verbose - acquireTokenSilent called`

**Comment by mikebper** [comment]
I followed that same documentation. I can get it to refresh every time from login as well, as opposed to just starting after the first refresh. I guess the accurate description of the issue is that the after the first time it needs to get a token from the token endpoint for a protected resource call (API), it then gets a new token from the token endpoint for every subsequent request, despite it not being necessary. It also doesn't seem to cache the newly acquired tokens.

For me that's after the initial token expiration. It doesn't need to call the token endpoint until then, because I get the most commonly used scope as part of my auth request. If I remove that scope from the auth request, however, then it has to call the token endpoint the first time an API call is made (as opposed to not until the token has expired), and from that point, every subsequent request it gets new tokens as well.

Does the issue only occur for you with a social login setup? In my case I have a separate AAD set up as a social login option using Open ID per the [B2C documentation (I did this exactly)](https://learn.microsoft.com/en-us/azure/active-directory-b2c/identity-provider-azure-ad-single-tenant?pivots=b2c-custom-policy).

For a local B2C account it works fine for me. I don't understand why the behavior is different for the AAD user, as I don't see any way the client could possibly even distinguish between the two. The token responses are identical in form (obviously the actual tokens and expiration are different). I also don't know how it could be B2C, as the tokens B2C is issuing look perfectly fine.

**Comment by SamP98** [comment]
For us it doesn't even work for local accounts. I have also tried downgrading @azure/msal-angular to 2.3.0 and the issue still occurs.

**Comment by SamP98** [comment]
I've found a solution. It was related to our scope.

I was going to just add the Authorization header manually in my own interceptor, but aquireTokenSilent was returning a blank string for the AccessToken.

After some searching, this solution worked for me. https://shawinnes.com/azure-b2c-and-msal/

He sets the scope in his requests equal to the b2c client id used in the msalConfig. Doing so gives you the correct AccessToken. So for my app, I set the scope in the protectedResourceMap to be the client id. This has solved the msal interceptor issue for me.

Hope this helps!

**Comment by mikebper** [comment]
So using client ID as scope for the auth request? I tried that and the issue I described still happens. My app works without client ID as a scope for the auth request. I see that MSAL automatically adds the scopes 'openid', 'profile', and 'email', and I specify my app's Read scope, so it doesn't have to get a new token as soon as the user requests data (which the associated endpoints are secured by my Read scope). Maybe you have to use client ID if you haven't created any scopes for the app in B2C. I have created and am using Read and Write scopes in B2C for my app.

**Comment by SamP98** [comment]
Yes, using the b2c client ID as scope fixed aquireTokenSilent.

What fixed the interceptor was putting it inside the protectedResourceMap inside the MSALInterceptorConfig.

The only thing that confuses/concerns me, is why is the MSALInterceptor calling aquireTokenSilent even if there is a valid AccessToken in the cache? What I've found seems like more of a workaround rather than a final solution.

Anyways, here's how I have all our configs.

```
const msalConfig: Configuration = {
    auth: {
        clientId: B2CClientID
        authority: https://MyOrg.b2clogin.com/MyOrg.onmicrosoft.com/OurPolicy,
        knownAuthorities: MyOrg.b2clogin.com,
        redirectUri: 'OurHomepage', 
        postLogoutRedirectUri: 'OurHomepage'
    },
    cache: {
        cacheLocation: BrowserCacheLocation.LocalStorage,
        storeAuthStateInCookie: isIE, 
    },
    system: {
        loggerOptions: {
          loggerCallback,
          logLevel: LogLevel.Trace,
          piiLoggingEnabled: true
        }
    }
}

export function MSALGuardConfigFactory(): MsalGuardConfiguration {
  return { 
    interactionType: InteractionType.Redirect,
    authRequest: {
      scopes: [B2CClientID, 'openid']
    }
  };
}

export function MSALInterceptorConfigFactory(): MsalInterceptorConfiguration {
  return {
    interactionType: InteractionType.Redirect,
    protectedResourceMap: new Map([
      [OurAPIEndpoint, B2CClientID] <------ this is what fixed it for me
    ])
  };
}
```

**Comment by mikebper** [comment]
I think it's normal for acquireTokenSilent to be called when retrieving tokens from cache. That is the method which does that activity. In absence of it being available from cache, it then gets a new token from the Identity Provider. It should only actually be reaching out to the Identity Provider at initial login, if it needs a token with a scope it doesn't already have one for, and when tokens are refreshed. When any of those happen, the associated network request to the IDP's token endpoint can be seen in dev tools.

I don't think your solution is relevant for me. I already have my scopes defined for each endpoint, like shown below.

```typescript
// app.module.ts -- MsalInterceptorConfiguration
{
  interactionType: InteractionType.Redirect,
  protectedResourceMap: new Map([
    [protectedResource.myEndpoint1.endpoint, protectedResource.myEndpoint1.scopes],
    [protectedResource.myEndpoint2.endpoint, protectedResource.myEndpoint2.scopes],
    [protectedResource.myEndpoint3.endpoint, protectedResource.myEndpoint3.scopes],
    [protectedResource.myEndpoint4.endpoint, protectedResource.myEndpoint4.scopes],
    [protectedResource.myEndpoint5.endpoint, protectedResource.myEndpoint5.scopes]
  ])
}
``` /

```typescript
// auth-config.ts -- constants used in MsalInterceptorConfiguration
export const protectedResource = {
    authRequest: {
        scopes: ['https://myTentant.onmicrosoft.com/<myClientId>/Read']
    },
    myEndpoint1: {
        endpoint: `${environment.myApiBaseUrl}/myEndpoint1`,
        scopes: ['https://myTentant.onmicrosoft.com/<myClientId>/Read']
    },
    myEndpoint2: {
        endpoint: `${environment.myApiBaseUrl}/myEndpoint2`,
        scopes: ['https://myTentant.onmicrosoft.com/<myClientId>/Read']
    },
    myEndpoint3: {
        endpoint: `${environment.myApiBaseUrl}/myEndpoint3`,
        scopes: ['https://myTentant.onmicrosoft.com/<myClientId>/Read']
    },
    myEndpoint4: {
        endpoint: `${environment.myApiBaseUrl}/myEndpoint4`,
        scopes: ['https://myTentant.onmicrosoft.com/<myClientId>/Read']
    },
    myEndpoint5: {
        endpoint: `${environment.myApiBaseUrl}/myEndpoint5`,
        scopes: ['https://myTentant.onmicrosoft.com/<myClientId>/Write']
    }
}
```

**Comment by sameerag** [comment]
@mikebper Will you please share you msal logs (preferably in verbose mode) - instructions [here](https://github.com/AzureAD/microsoft-authentication-library-for-js/blob/dev/lib/msal-angular/docs/v2-docs/logging.md) and also any network trace (fiddler)?

You can send then network trace to my [email](https://github.com/sameerag).

**Comment by mikebper** [comment]
Fortunately, it looks like I won't have to--the problem has been resolved!  When collecting the logs for you, I noticed the following warning message. I'm not sure how I overlooked it before:

> auth-config.ts:42 [Tue, 25 Apr 2023 23:00:30 GMT] : [<GUID>] : @azure/msal-common@8.0.0 : Warning - Account used to refresh tokens not in persistence, refreshed tokens will not be stored in the cache

Doing a search on that error immediately led me to the solution. The problem was that the "tenantId" claim was missing on the refresh token for external social login AAD users due to it not being specified in the custom policy on the "RefreshTokenReadAndSetup" TechnicalProfile (used by the refresh token user journey). Adding the claim there resolved the problem:
```xml
<OutputClaim ClaimTypeReferenceId="tenantId"/>
``` 

The claim was present for local B2C users, but not for external users--I'm not exactly sure why. I had recently noticed the only difference in the token response was absence of "tid", but I didn't think that could be the problem at the time. Apparently it was! I don't know why MSAL won't store the token in cache just because the tid claim was missing--it's still a valid token which passes validation in both API management and Microsoft Identity platform in the API.

Anyway, it's working properly now for both local B2C users and social login/external AAD users. Thanks!
